MewsSystems/developers
# Movie Search App

Application has been deployed via Vercel - https://developers-frontend-jake-towers-projects.vercel.app

## What did you change?

* Added a Search view containing a search bar to search for movies in the MovieDB
  * The search component makes a search automatically after typing into the input is finished by using a debounce of 500ms 
  * When the results have loaded, you can click on one to visit the Movie Details view for it
  * There is also pagination for going through more results for you search query
* Added a Movie Details view for viewing additional details about a selected movie from the MovieDB
* Added a Pull Request template to this project to create a template like this when making a PR
* Added a GitHub Actions workflow to deploy the application to Vercel
* Added husky to create a pre-commit hook to ensure files are linted before committing
* Added `.nvmrc` file to inform the project developers the appropriate version of Node to use
* Added prettier to format the code
* Added Jest for unit testing
* Added Playwright for E2E testing
* Used Next/Image to help with image loading
  * Used a base64 image to show a placeholder while they load 
* Added Styled Components for styling
  * Additionally created a theme and used this so it's easier to replace colours and create alternative themes in the future 
* Added some comments in the code to explain decisions or thoughts I had at that time. I wouldn't normally do this but it's useful for the purposes of this task

## Are any updates required?

* You will need to run `npm i` to install the packages
* There are new environment variables that have been added

## How have you tested it?

* Unit tests are all passing
* E2E tests are all passing
* Manually tested that:
  * When there is no search term, you see a prompt encouraging you to search
  * When searching, results are returned relevant to the search term 
    * These results can be selected
    * These results can be paginated through
  * When no results are returned, the user is informed
  * The user is able to navigate back from the movie details page to the results page 
  * When there is no response from the API the application responds appropriately
  * The application displays appropriately across desktop/tablet/mobile views
  * The application displays appropriately across browsers

## Screenshots

### Desktop

#### Search View 

![Screenshot 2024-05-05 at 16 11 51](https://github.com/MewsSystems/developers/assets/10836097/c4a75595-6d77-49ce-b413-a6f30da47832)

![Screenshot 2024-05-05 at 16 12 13](https://github.com/MewsSystems/developers/assets/10836097/01d92258-37f7-4564-9164-bcc35673e53f)

![Screenshot 2024-05-05 at 16 13 52](https://github.com/MewsSystems/developers/assets/10836097/7120c177-2fc0-40d7-9caa-573f772a923e)

#### Movie Details View

![Screenshot 2024-05-05 at 16 12 31](https://github.com/MewsSystems/developers/assets/10836097/f4b32a6f-2f94-4fbc-b671-a6c93e1f9af8)

### Mobile

#### Search View 

![Screenshot 2024-05-05 at 16 14 33](https://github.com/MewsSystems/developers/assets/10836097/7b5b5937-889d-46d0-884d-416f72cc3f83)

![Screenshot 2024-05-05 at 16 14 41](https://github.com/MewsSystems/developers/assets/10836097/63984bbb-4962-43c4-8598-87031a25986f)

![Screenshot 2024-05-05 at 16 14 22](https://github.com/MewsSystems/developers/assets/10836097/6a5e8087-23ec-4678-b30b-31d176a4d838)

#### Movie Details View

![Screenshot 2024-05-05 at 16 14 45](https://github.com/MewsSystems/developers/assets/10836097/658b5580-62e5-4829-8845-277a3a64af5e)

### Mock Designs

I created a Miro board to mock up some designs, ideas, and inspiration 

![Screenshot 2024-05-05 at 16 16 31](https://github.com/MewsSystems/developers/assets/10836097/ecdb7e39-9495-4cdb-9228-4857125b81e4)
